### PR TITLE
Search NULL values on all field types

### PR DESCRIPTION
**Pull request by michael-mri**

### CODE REVIEW

**Comment by michael-mri** [comment]
Few things to note:

- new setting NULL_VALUE which is the default null value. NULL_CHOICE_VALUE defaults to it.
- haven't yet implemented tests for each type of field
- haven't yet updated the documentation
- we may want to add null value in forms some way (especially for ModelChoiceFilter), although I am not sure how....

I struggled a bit on making form validation work: see filterset.py, lines 222-227. 

Comments and suggestions are welcome :)

**Comment by rpkilby** [comment]
I can provide more details, but I'm kind of -1 on these changes since it will play poorly with non-text inputs. My recommendation is to use a separate filter to handle null values explicitly, as demonstrated in the [docs](https://django-filter.readthedocs.io/en/latest/guide/tips.html#filtering-by-null-values).

An example of poor form usage would be number inputs. The html number input does not allow text entry, so it's not possible to submit the 'null' string. If you manually coerce the querystring to filter by the null string, then the form field will have to be altered to a text input. When re-rendering the form, the number controls will not be present. This also applies to date, datetime, and custom fields (such as any `MultiValueField`s).

**Comment by michael-mri** [comment]
Indeed I agree they play poorly when it comes to use them in inputs, especially those generated automatically on DRF web interface for example.
However this make REST interface more friendly, being able to filter for example author=1, author=2 or author=null the same way. 

I think this also makes sense for choices filters (ChoiceFilter, ModelChoiceFilter, MultipleChoiceFilter and ModelMultipleChoiceFilter) because these are rendered using input like selectboxes where we can define what null value is.
Maybe null_value should be restricted to those filters ?

**Comment by rpkilby** [comment]
> I think this also makes sense for choices filters (`ChoiceFilter`, `ModelChoiceFilter`, `MultipleChoiceFilter` and `ModelMultipleChoiceFilter`) because these are rendered using input like selectboxes where we can define what null value is.

Setting [`NULL_CHOICE_LABEL`](http://django-filter.readthedocs.io/en/develop/ref/settings.html?highlight=NULL_CHOICE_LABEL#filters-null-choice-label) to a non-`None` value should enable null value filtering for both `ChoiceFilter` and `ModelChoiceFilter`. 

If I recall correctly, null-choice filtering was not added to `MultipleChoiceFilter` since it would be generally non-sensical. eg, I want to filter all blogs that have no entries and that have this specific entry. You would end up with a contradictory query.

**Comment by carltongibson** [comment]
`all blogs that have no entries OR that have this specific entry` would make sense though right?

**Comment by rpkilby** [comment]
> `all blogs that have no entries OR that have this specific entry` would make sense though right?

True, and it would not be much effort to extend the existing functionality it to `MultipleChoiceFilter` and `ModelMultipleChoiceFilter`.

File: django_filters/filterset.py [hunk 1]
```diff
@@ -222,7 +226,7 @@ def form(self):
             fields = OrderedDict([
                 (name, filter_.field)
                 for name, filter_ in six.iteritems(self.filters)])
-
+
```
Comment (carltongibson) [inline django_filters/filterset.py]: Can you set your editor to trim trailing whitespace?

**Pull request closed without merging**